cirq/protocols/decompose_protocol.py

```diff
@@ -387,3 +382,66 @@ def _try_decompose_into_operations_and_qubits(
         return result, qubits, tuple(qid_shape_dict[q] for q in qubits)
 
     return None, (), ()
+
+
+def decompose_preserving_structure(
```

Comment:
Do we really need a completely new protocol for this? Given that decompose is challenging to understand/reason about already, I feel like introducing another version makes it even more challenging...we'll also need to keep the API exactly the same which is another smell...wouldn't another parameter to decompose() work?

Reply:
I'm not strictly opposed to folding this into `decompose` directly. Something like this?
```
def decompose(..., preserve_structure=False):
    if preserve_structure:
        if intercepting_decomposer is not None:
            return ValueError(...)
        return _decompose_preserving_structure(...)
    {...}
```

Reply:
Change applied.